docs(v1.12.0): update important notes (#1363)
longhorn/website: content/docs/1.12.0/important-notes/_index.md

  - [Encrypted Volume Size After Engine Upgrade](#encrypted-volume-size-after-engine-upgrade)
- [General](#general)
  - [Kubernetes Version Requirement](#kubernetes-version-requirement)
  - [Manual Checks Before Upgrade](#manual-checks-before-upgrade)
- [Scheduling](#scheduling)
  - [Topology-Aware PV Node Affinity Control](#topology-aware-pv-node-affinity-control)
- [Stability](#stability)
  - [Configurable Engine Image Pod Liveness Probe](#configurable-engine-image-pod-liveness-probe)
- [Resource Efficiency](#resource-efficiency)
  - [Longhorn Manager Memory Optimization](#longhorn-manager-memory-optimization)
- [Networking](#networking)
  - [Dual-Stack Cluster Support](#dual-stack-cluster-support)
- [Monitoring](#monitoring)
  - [Toggle Kubernetes Metrics Server Integration](#toggle-kubernetes-metrics-server-integration)
- [Command-Line Tool](#command-line-tool)
  - [On-Demand Snapshot Checksum Calculation](#on-demand-snapshot-checksum-calculation)
- [V2 Data Engine](#v2-data-engine)
  - [Fast Cloning](#fast-cloning)
  - [IPv6 Support](#ipv6-support)
  - [Longhorn System Upgrade](#longhorn-system-upgrade)
  - [Sharding (Experimental)](#sharding-experimental)

## Removal

> Longhorn v{{< current-version >}} removes V2 Backing Image. Use containerized data importer (CDI) to import images into Longhorn instead.

Reviewer: ```suggestion
V2 Backing Images are removed in Longhorn v{{< current-version >}}. Use the Containerized Data Importer (CDI) to import images into Longhorn for compatibility with the current engine
```

For more information, see [Issue #13087](https://github.com/longhorn/longhorn/issues/13087).

### Replica Rebuild Progress Reporting

Longhorn v{{< current-version >}} fixes a replica rebuild progress reporting bug that could display values greater than 100% after file-sync retries on unstable networks. Progress accounting is now reset correctly for retried files, so rebuild progress remains within the valid 0% to 100% range.

For more information, see [Issue #12949](https://github.com/longhorn/longhorn/issues/12949).

### Replica Auto-Balance Scheduling Loop

Longhorn v{{< current-version >}} fixes a regression in replica auto-balance that could trigger a repeated replica create-and-delete loop when `Replica Auto Balance` was set to `best-effort`. In affected clusters, Longhorn could keep scheduling an extra replica instead of stabilizing at the configured replica count.

For more information, see [Issue #12926](https://github.com/longhorn/longhorn/issues/12926).

### Replica CR Leak During Failed Local Scheduling

> Longhorn v{{< current-version >}} fixes a replica scheduling issue that could leak large numbers of stopped Replica CRs when `dataLocality` was set to `best-effort` and the node did not have enough eligible local disk space for another replica. In affected clusters, recurring reconciliation could keep creating placeholder Replica CRs instead of reusing a single failed-schedule placeholder.

Reviewer: ```suggestion
Longhorn v{{< current-version >}} fixes a replica scheduling issue where large numbers of stopped Replica CRs could accumulate when `dataLocality` was set to `best-effort` and the node did not have enough eligible local disk space for another replica. In affected clusters, recurring reconciliation could keep creating placeholder Replica CRs instead of reusing a single failed-schedule placeholder.
```

**After upgrading to v1.12**: Once the engine image is upgraded for an encrypted volume, Longhorn automatically expands the backend size by 16 MiB. The dm-crypt device then exposes the full requested size (e.g., exactly 1 GiB for a 1 GiB volume). Existing data is not affected.

**Live migration restriction**: Encrypted migratable volumes cannot be live-migrated when using an engine image with a CLI API version older than 12 (pre-v1.12 engine images). Upgrade the engine image to v1.12 or later before attempting live migration of encrypted volumes.

For more information, see [Issue #9205](https://github.com/longhorn/longhorn/issues/9205).

## General

### Kubernetes Version Requirement

Because the CSI external snapshotter is upgraded to v8.2.0, all clusters must be running Kubernetes v1.25 or later before upgrading to Longhorn v{{< current-version >}}.

### Manual Checks Before Upgrade

Automated pre-upgrade checks do not cover all scenarios. Manual checks via kubectl or the UI are recommended:

- Ensure all V2 Data Engine volumes are detached and replicas are stopped. The V2 engine does not support live upgrades.

> - Avoid upgrading when volumes are "Faulted", as unusable replicas may be deleted, causing permanent data loss if no backups exist.

Reviewer: ```suggestion
- Avoid upgrading when volumes are in the "Faulted" state, as unusable replicas may be deleted, causing permanent data loss if no backups exist.
```

For more information, see [Issue #12684](https://github.com/longhorn/longhorn/issues/12684) and [Topology-Aware Provisioning](../nodes-and-volumes/nodes/topology-aware-provisioning).

## Stability

### Configurable Engine Image Pod Liveness Probe

Longhorn v{{< current-version >}} adds settings to configure the engine-image DaemonSet liveness probe period, timeout, and failure threshold. These settings help reduce unnecessary engine-image pod restarts on resource-constrained clusters, especially during upgrades or transient CPU spikes.

For more information, see [Issue #12846](https://github.com/longhorn/longhorn/issues/12846) and [Settings](../references/settings#engine-image-pod-liveness-probe-period).

## Resource Efficiency

### Longhorn Manager Memory Optimization

Longhorn v{{< current-version >}} optimizes longhorn-manager informer caching to reduce memory usage, especially in large clusters with high pod counts. This lowers cluster-wide memory overhead caused by repeated caching of non-Longhorn pod data on every manager instance.

For more information, see [Issue #12771](https://github.com/longhorn/longhorn/issues/12771).

## Networking

### Dual-Stack Cluster Support

> Longhorn supports dual-stack Kubernetes clusters under a limited condition: all nodes must be configured with their IP families in the same order (either all IPv4-first, or all IPv6-first). When the order is consistent, Longhorn uses the first IP family of each node and operates correctly. This applies to both the V1 and V2 data engines.

Reviewer: ```suggestion
Longhorn supports dual-stack Kubernetes clusters under a specific requirement: all nodes must be configured with their IP families in the same order (either all IPv4-first, or all IPv6-first). When the order is consistent, Longhorn uses the first IP family of each node and operates correctly. This applies to both the V1 and V2 data engines.
```
